### PR TITLE
Started the exercise.

### PR DESCRIPTION
**Pull request by janlemcke**
1. Create chunk like data structure
2. Flatten this 2D data structure into 1D array.
3. Create & write file with flattened data.

TODO:
Reading + Validating.

### CODE REVIEW

File: src/bin/playground.cpp [hunk 1]
```diff
@@ -1,10 +1,230 @@
...
+std::string fail_and_close_file(int32_t fd, std::string message, int error_num) {
+  close(fd);
+  return message + std::strerror(error_num);
+}
+
+void printData(chunk_type data) {
+  const ssize_t ROW_COUNT = data.size();
+  const ssize_t COLUMN_COUNT = data.at(0)->size();
+  std::cout << "Data:" << std::endl;
+  for (ssize_t i = 0; i < COLUMN_COUNT; i++) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:23]: auto (from now on: [1])

File: src/bin/playground.cpp [hunk 2]
```diff
@@ -1,10 +1,230 @@
...
+    for (ssize_t j = 0; j < ROW_COUNT; j++) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:24]: [1]

File: src/bin/playground.cpp [hunk 3]
```diff
@@ -1,10 +1,230 @@
...
+      std::cout << data.at(j)->at(i) << " ";
+    }
+    std::cout << std::endl;
+  }
+}
+
+void printVector(std::vector<int32_t> vector) {
+  std::cout << "Vector:" << std::endl;
+  for (int32_t i : vector) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:33]: index instead of i
[1]

File: src/bin/playground.cpp [hunk 4]
```diff
@@ -1,10 +1,230 @@
...
+    std::cout << i << ' ';
+  }
+  std::cout << std::endl;
+}
+
+std::vector<int32_t> flatten(chunk_type const& chunk) {
+  const ssize_t ROW_COUNT = chunk.size();
+  const ssize_t COLUMN_COUNT = chunk.at(0)->size();
+
+  std::vector<int32_t> flattened;
```
Comment (SanJSp) [inline src/bin/playground.cpp:43]: [1]

File: src/bin/playground.cpp [hunk 5]
```diff
@@ -1,10 +1,230 @@
...
+  flattened.reserve(ROW_COUNT * COLUMN_COUNT);
+
+  for (ssize_t i = 0; i < COLUMN_COUNT; i++) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:46]: [1]

File: src/bin/playground.cpp [hunk 6]
```diff
@@ -1,10 +1,230 @@
...
+    for (ssize_t j = 0; j < ROW_COUNT; j++) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:47]: [1]

File: src/bin/playground.cpp [hunk 7]
```diff
@@ -1,10 +1,230 @@
...
+      flattened.push_back(chunk.at(j)->at(i));
+    }
+  }
+  return flattened;
+}
+
+chunk_type createChunk(const uint32_t row_count, const uint32_t column_count) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:54]: snake case

File: src/bin/playground.cpp [hunk 8]
```diff
@@ -1,10 +1,230 @@
...
+  auto chunk = chunk_type{};
+  const auto VALUE_AMOUNT = column_count * row_count;
+
+  std::cout << "We want to create one chunk with " << column_count << " columns, " << row_count << " rows and thus "
+            << VALUE_AMOUNT << " values." << std::endl;
+  chunk.reserve(column_count);
+
+  // create columns
+  for (uint32_t i = 0; i < column_count; ++i) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:63]: [1]

File: src/bin/playground.cpp [hunk 9]
```diff
@@ -1,10 +1,230 @@
...
+    auto new_column = std::vector<int32_t>{};
+    new_column.reserve(row_count);
+    chunk.push_back(std::make_shared<std::vector<int32_t>>(new_column));
+  }
+
+  std::cout << chunk.size() << " columns created." << std::endl;
+
+  // create rows / insert values
+  for (uint32_t i = 0; i < VALUE_AMOUNT; ++i) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:72]: [1], index

File: src/bin/playground.cpp [hunk 10]
```diff
@@ -1,10 +1,230 @@
...
+    auto column_index = i % column_count;
+    auto column = chunk.at(column_index);
+    column->push_back(i * column_index);
+  }
+
+  std::cout << "Values populated." << std::endl;
+  return chunk;
+}
+
+void writeDataToFile(std::vector<int32_t> flattened_chunk, const char* filename) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:82]: snake case

File: src/bin/playground.cpp [hunk 11]
```diff
@@ -1,10 +1,230 @@
...
+  const ssize_t NUMBER_OF_BYTES = flattened_chunk.size() * sizeof(int32_t);
+  auto fd = int32_t{};
+  Assert(((fd = creat(filename, O_WRONLY)) >= 1), fail_and_close_file(fd, "Create error: ", errno));
+  chmod(filename, S_IRWXU);  // enables owner to rwx file
+  Assert((write(fd, std::data(flattened_chunk), NUMBER_OF_BYTES) == NUMBER_OF_BYTES),
+         fail_and_close_file(fd, "Write error: ", errno));
+  close(fd);
+}
+
+std::vector<int32_t> readDataFromFile(size_t number_of_bytes, const char* filename) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:92]: snake case

File: src/bin/playground.cpp [hunk 12]
```diff
@@ -1,10 +1,230 @@
...
+  for (auto index = size_t{0}; index < NUMBER_OF_ITEMS; ++index) {
+    read_data.push_back(map[index]);
+  }
+
+  // Remove memory mapping after job is done.
+  Assert((munmap(map, number_of_bytes) == 0), fail_and_close_file(fd, "Unmapping failed: ", errno));
+  return read_data;
+}
+
+chunk_type recreateChunk(const uint32_t row_count, const uint32_t column_count, std::vector<int32_t> data_vector) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:114]: snake case

**Review by SanJSp — CHANGES_REQUESTED** [review]
Minor refactorings. Great job!

**Comment by SanJSp** [comment]
Please test this again, implemented my feedback

File: src/bin/playground.cpp [hunk 13]
```diff
@@ -1,10 +1,230 @@
...
+    auto column_index = index % column_count;
+    auto column = chunk.at(column_index);
+    column->push_back(index * column_index);
+  }
+
+  std::cout << "Values populated." << std::endl;
+  return chunk;
+}
+
+void write_data_to_file(std::vector<int32_t> flattened_chunk, const char* filename) {
```
[the remark this answers is not in the document]
Comment (tjjordan) [reply src/bin/playground.cpp:82]: Also filename should be a reference to a string.

File: src/bin/playground.cpp [hunk 14]
```diff
@@ -1,10 +1,230 @@
...
+  }
+}
+
+int calculate_sum_across_column(chunk_type chunk, u_int32_t col_index) {
+  auto column = chunk.at(col_index);
+  auto sum = std::accumulate(column->begin(), column->end(), uint64_t{0});
+  return sum;
+}
+
+void print_row(chunk_type chunk, u_int32_t row_index) {
```
Comment (tjjordan) [inline src/bin/playground.cpp:197]:
```suggestion
void print_row(const chunk_type& chunk, const u_int32_t row_index) {
```

File: src/bin/playground.cpp [hunk 15]
```diff
@@ -1,10 +1,230 @@
...
+  const auto NUMBER_OF_BYTES = static_cast<ssize_t>(flattened_chunk.size() * sizeof(int32_t));
+  auto fd = int32_t{};
+  Assert(((fd = creat(filename, O_WRONLY)) >= 1), fail_and_close_file(fd, "Create error: ", errno));
+  chmod(filename, S_IRWXU);  // enables owner to rwx file
+  Assert((write(fd, std::data(flattened_chunk), NUMBER_OF_BYTES) == NUMBER_OF_BYTES),
+         fail_and_close_file(fd, "Write error: ", errno));
+  close(fd);
+}
+
+std::vector<int32_t> read_data_from_file(size_t number_of_bytes, const char* filename) {
```
Comment (tjjordan) [inline src/bin/playground.cpp:92]:
```suggestion
std::vector<int32_t> read_data_from_file(const size_t number_of_bytes, const char* filename) {
```

File: src/bin/playground.cpp [hunk 16]
```diff
@@ -1,10 +1,230 @@
...
+  auto read_data = std::vector<int32_t>{};
+  const auto NUMBER_OF_ITEMS = number_of_bytes / sizeof(int32_t);
```
Comment (tjjordan) [inline src/bin/playground.cpp:94]: lower case

File: src/bin/playground.cpp [hunk 17]
```diff
@@ -1,10 +1,230 @@
...
+  auto* map = reinterpret_cast<int32_t*>(mmap(NULL, number_of_bytes, PROT_READ, MAP_PRIVATE, fd, OFFSET));
+  Assert((map != MAP_FAILED), fail_and_close_file(fd, "Mapping Failed: ", errno));
+
+  madvise(map, number_of_bytes, MADV_SEQUENTIAL);
+
+  auto chunk = chunk_type{};
+  chunk.reserve(column_count);
+
+  for (auto index = size_t{0}; index < NUMBER_OF_ITEMS; ++index) {
+    auto column_index = index % column_count;
```
Comment (tjjordan) [inline src/bin/playground.cpp:159]:
```suggestion
    const auto column_index = index % column_count;
```

File: src/bin/playground.cpp [hunk 18]
```diff
@@ -1,10 +1,230 @@
...
+    if (column_index < chunk.size()) {
+      auto column = chunk.at(column_index);
+      column->push_back(map[index]);
```
Comment (tjjordan) [inline src/bin/playground.cpp:162]: Prefer emplace_back

File: src/bin/playground.cpp [hunk 19]
```diff
@@ -1,10 +1,230 @@
...
+  const auto ROW_COUNT = original_chunk.size();
+  const auto COLUMN_COUNT = original_chunk.at(0)->size();
+
+  Assert(ROW_COUNT == read_chunk.size(), "Amount of rows doesn't match!");
+  Assert(COLUMN_COUNT == read_chunk.at(0)->size(), "Amount of columns doesn't match!");
+
+  for (auto column_index = size_t{0}; column_index < COLUMN_COUNT; column_index++) {
+    for (auto row_index = size_t{0}; row_index < ROW_COUNT; row_index++) {
+      auto original_value = original_chunk.at(row_index)->at(column_index);
+      auto read_value = read_chunk.at(row_index)->at(column_index);
```
Comment (tjjordan) [inline src/bin/playground.cpp:185]: both can be const

File: src/bin/playground.cpp [hunk 20]
```diff
@@ -1,10 +1,230 @@
...
+  }
+
+  // Remove memory mapping after job is done.
+  Assert((munmap(map, number_of_bytes) == 0), fail_and_close_file(fd, "Unmapping failed: ", errno));
+  return chunk;
+}
+
+void sanity_check(chunk_type original_chunk, chunk_type read_chunk) {
+  const auto ROW_COUNT = original_chunk.size();
+  const auto COLUMN_COUNT = original_chunk.at(0)->size();
```
Comment (tjjordan) [inline src/bin/playground.cpp:177]: lower_case

File: src/bin/playground.cpp [hunk 21]
```diff
@@ -1,10 +1,230 @@
...
+      chunk.push_back(std::make_shared<std::vector<int32_t>>(new_column));
+    }
+  }
+
+  // Remove memory mapping after job is done.
+  Assert((munmap(map, number_of_bytes) == 0), fail_and_close_file(fd, "Unmapping failed: ", errno));
+  return chunk;
+}
+
+void sanity_check(chunk_type original_chunk, chunk_type read_chunk) {
```
Comment (tjjordan) [inline src/bin/playground.cpp:175]: Both parameters should be const and references.

File: src/bin/playground.cpp [hunk 22]
```diff
@@ -1,10 +1,230 @@
...
+  for (auto column_index = size_t{0}; column_index < COLUMN_COUNT; column_index++) {
+    for (auto row_index = size_t{0}; row_index < ROW_COUNT; row_index++) {
+      auto original_value = original_chunk.at(row_index)->at(column_index);
+      auto read_value = read_chunk.at(row_index)->at(column_index);
+      Assert(original_value == read_value, "Values are not the same!");
+    }
+  }
+}
+
+int calculate_sum_across_column(chunk_type chunk, u_int32_t col_index) {
```
Comment (tjjordan) [inline src/bin/playground.cpp:191]: Both parameters should be const, chunk_type sould be a reference.

Comment (tjjordan) [reply src/bin/playground.cpp:191]: Avoid using int!
Use size_t, uint32_t or uint64_t instead.

File: src/bin/playground.cpp [hunk 23]
```diff
@@ -1,10 +1,230 @@
...
+  auto column = chunk.at(col_index);
+  auto sum = std::accumulate(column->begin(), column->end(), uint64_t{0});
```
Comment (tjjordan) [inline src/bin/playground.cpp:193]: both can be const

File: src/bin/playground.cpp [hunk 24]
```diff
@@ -1,10 +1,230 @@
...
+  for (auto index = ssize_t{0}; index < COLUMN_COUNT; ++index) {
+    std::cout << chunk.at(index)->at(row_index) << " ";
+  }
+  std::cout << std::endl;
+}
+
 int main() {
-  std::cout << "Hello world!!" << std::endl;
+  std::cout << "Playground started." << std::endl;
+  const char* filename = "flattened_vector.txt";
```
Comment (tjjordan) [inline src/bin/playground.cpp:208]: std::string?

File: src/bin/playground.cpp [hunk 25]
```diff
@@ -1,10 +1,230 @@
...
+  const auto COLUMN_COUNT = uint32_t{23};
+  const auto ROW_COUNT = uint32_t{65000};
+  auto chunk = create_chunk(ROW_COUNT, COLUMN_COUNT);
```
Comment (tjjordan) [inline src/bin/playground.cpp:211]: const?

File: src/bin/playground.cpp [hunk 26]
```diff
@@ -1,10 +1,230 @@
...
+
+  std::cout << "Finished writing." << std::endl;
+  std::cout << "Start reading." << std::endl;
+
+  /*
+  auto read_vector = readDataFromFile(NUMBER_OF_BYTES, filename);
+  printVector(read_vector);
+  auto read_chunk = recreateChunk(ROW_COUNT, COLUMN_COUNT, read_vector);
+   or */
+  auto read_chunk = read_data_from_file_as_chunk(COLUMN_COUNT, NUMBER_OF_BYTES, filename);
```
Comment (tjjordan) [inline src/bin/playground.cpp:224]: const

File: src/bin/playground.cpp [hunk 27]
```diff
@@ -1,10 +1,230 @@
...
+  sanity_check(chunk, read_chunk);
+  auto sum = calculate_sum_across_column(read_chunk, 17);
```
Comment (tjjordan) [inline src/bin/playground.cpp:226]: const

Comment (tjjordan) [inline src/bin/playground.cpp:226]: const

File: src/bin/playground.cpp [hunk 28]
```diff
@@ -1,10 +1,230 @@
...
+}
+
 int main() {
-  std::cout << "Hello world!!" << std::endl;
+  std::cout << "Playground started." << std::endl;
+  const char* filename = "flattened_vector.txt";
+  const auto COLUMN_COUNT = uint32_t{23};
+  const auto ROW_COUNT = uint32_t{65000};
+  auto chunk = create_chunk(ROW_COUNT, COLUMN_COUNT);
+  auto flattened_chunk = flatten(chunk);
```
Comment (tjjordan) [inline src/bin/playground.cpp:212]: const

File: src/bin/playground.cpp [hunk 29]
```diff
@@ -1,10 +1,230 @@
...
+  const ssize_t NUMBER_OF_BYTES = flattened_chunk.size() * sizeof(int32_t);
```
Comment (tjjordan) [inline src/bin/playground.cpp:213]: auto

**Review by tjjordan — CHANGES_REQUESTED** [review]
Functionality looks good.
But some (very picky) style guide remarks.

File: src/bin/playground.cpp [hunk 30]
```diff
@@ -1,10 +1,230 @@
...
+         fail_and_close_file(fd, "Write error: ", errno));
+  close(fd);
+}
+
+std::vector<int32_t> read_data_from_file(size_t number_of_bytes, const char* filename) {
+  auto read_data = std::vector<int32_t>{};
+  const auto NUMBER_OF_ITEMS = number_of_bytes / sizeof(int32_t);
+  read_data.reserve(NUMBER_OF_ITEMS);
+  auto fd = int32_t{};
+  Assert(((fd = open(filename, O_RDONLY)) >= 0), fail_and_close_file(fd, "Open error: ", errno));
```
Comment (tjjordan) [inline src/bin/playground.cpp:97]:
```suggestion
  Assert((fd = open(filename, O_RDONLY)) >= 0, fail_and_close_file(fd, "Open error: ", errno));
```

File: src/bin/playground.cpp [hunk 31]
```diff
@@ -1,10 +1,230 @@
...
+
+  // Getting the mapping to memory.
+  const auto OFFSET = off_t{0};
+  auto* map = reinterpret_cast<int32_t*>(mmap(NULL, number_of_bytes, PROT_READ, MAP_PRIVATE, fd, OFFSET));
+  Assert((map != MAP_FAILED), fail_and_close_file(fd, "Mapping Failed: ", errno));
+
+  madvise(map, number_of_bytes, MADV_SEQUENTIAL);
+  for (auto index = size_t{0}; index < NUMBER_OF_ITEMS; ++index) {
+    read_data.push_back(map[index]);
```
Comment (tjjordan) [inline src/bin/playground.cpp:106]: emplace_back

File: src/bin/playground.cpp [hunk 32]
```diff
@@ -1,10 +1,230 @@
...
+  }
+
+  // Remove memory mapping after job is done.
+  Assert((munmap(map, number_of_bytes) == 0), fail_and_close_file(fd, "Unmapping failed: ", errno));
+  return read_data;
+}
+
+chunk_type recreate_chunk(const uint32_t row_count, const uint32_t column_count, std::vector<int32_t> data_vector) {
```
Comment (tjjordan) [inline src/bin/playground.cpp:114]: data_vector should be a reference

File: src/bin/playground.cpp [hunk 33]
```diff
@@ -1,10 +1,230 @@
...
+  Assert(row_count % column_count == 0, "Row count is not a multiple of column count!");
```
Comment (tjjordan) [inline src/bin/playground.cpp:115]: Why must the row_count be a multiple of the column count?

File: src/bin/playground.cpp [hunk 34]
```diff
@@ -1,10 +1,230 @@
...
+  auto chunk = chunk_type{};
+  const auto VALUE_AMOUNT = data_vector.size();
```
Comment (tjjordan) [inline src/bin/playground.cpp:117]: lower case

File: src/bin/playground.cpp [hunk 35]
```diff
@@ -1,10 +1,230 @@
...
+    auto column_index = index % column_count;
+    auto column = chunk.at(column_index);
+    column->push_back(data_vector.at(index));
+  }
+
+  std::cout << "Values loaded." << std::endl;
+  return chunk;
+}
+
+chunk_type read_data_from_file_as_chunk(const uint32_t column_count, size_t number_of_bytes, const char* filename) {
```
Comment (tjjordan) [inline src/bin/playground.cpp:143]: number_of_bytes can be const
filename should be a std::string

File: src/bin/playground.cpp [hunk 36]
```diff
@@ -1,10 +1,230 @@
...
+  const auto NUMBER_OF_ITEMS = number_of_bytes / sizeof(int32_t);
```
Comment (tjjordan) [inline src/bin/playground.cpp:144]: lower case

File: src/bin/playground.cpp [hunk 37]
```diff
@@ -1,10 +1,230 @@
...
+  auto fd = int32_t{};
+  Assert(((fd = open(filename, O_RDONLY)) >= 0), fail_and_close_file(fd, "Open error: ", errno));
```
Comment (tjjordan) [inline src/bin/playground.cpp:146]:
```suggestion
  Assert((fd = open(filename, O_RDONLY)) >= 0, fail_and_close_file(fd, "Open error: ", errno));
```

File: src/bin/playground.cpp [hunk 38]
```diff
@@ -1,10 +1,230 @@
...
+  }
+
+  std::cout << "Values populated." << std::endl;
+  return chunk;
+}
+
+void write_data_to_file(std::vector<int32_t> flattened_chunk, const char* filename) {
+  const auto NUMBER_OF_BYTES = static_cast<ssize_t>(flattened_chunk.size() * sizeof(int32_t));
+  auto fd = int32_t{};
+  Assert(((fd = creat(filename, O_WRONLY)) >= 1), fail_and_close_file(fd, "Create error: ", errno));
```
Comment (tjjordan) [inline src/bin/playground.cpp:85]:
```suggestion
  Assert((fd = creat(filename, O_WRONLY)) >= 1, fail_and_close_file(fd, "Create error: ", errno));
```

**Review by tjjordan — CHANGES_REQUESTED** [review]
Small addition.

**Pull request closed without merging**